src/isc-dhcp/patch/0009-Support-for-dual-tor-scenario.patch

```diff
@@ -0,0 +1,308 @@
++						i = -1;
++						break;
++					}
++				}
++
++				if (i == -1) {
+ 					break;
++				} else {
++					out = NULL;
+ 				}
+ 			}
+-
+-			if (i == -1)
+-				break;
++		} else {
++			out = NULL;
+ 		}
+ 	} else {
+-		out = NULL;
++		/* Find the interface that corresponds to the giaddr in the packet. */
++		if (packet->giaddr.s_addr) {
++			for (out = interfaces; out; out = out->next) {
++				int i;
++
++				for (i = 0 ; i < out->address_count ; i++ ) {
++					if (out->addresses[i].s_addr ==
++						packet->giaddr.s_addr) {
++						i = -1;
++						break;
++					}
++				}
++
++				if (i == -1)
++					break;
++			}
++		} else {
++			out = NULL;
++		}
+ 	}
+ 
+ 	/* If it's a bootreply, forward it to the client. */
+@@ -913,6 +962,10 @@ do_relay4(struct interface_info *ip, struct dhcp_packet *packet,
+ 			return;
+ 		}
+ 
++		/* This bootreply does not belong to the current vlan. */
++		if (enable_support_for_dual_tor && !out)
++			return;
++
+ 		if (!(packet->flags & htons(BOOTP_BROADCAST)) &&
+ 			can_unicast_without_arp(out)) {
+ 			to.sin_addr = packet->yiaddr;
+@@ -945,9 +998,13 @@ do_relay4(struct interface_info *ip, struct dhcp_packet *packet,
+ 			return;
+ 
+ 		if (!out) {
+-			log_error("Packet to bogus giaddr %s.\n",
+-			      inet_ntoa(packet->giaddr));
+-			++bogus_giaddr_drops;
++			if (!enable_support_for_dual_tor) {
++				log_error("Packet to bogus giaddr %s.\n", inet_ntoa(packet->giaddr));
++				++bogus_giaddr_drops;
++			} else {
++				log_error("Packet to bogus yiaddr %s.\n", inet_ntoa(packet->yiaddr));
++				++bogus_yiaddr_drops;
++			}
+ 			return;
+ 		}
+ 
+@@ -987,8 +1044,14 @@ do_relay4(struct interface_info *ip, struct dhcp_packet *packet,
+ 	   forward the response to the correct net.    If it's already
+ 	   set, the response will be sent directly to the relay agent
+ 	   that set giaddr, so we won't see it. */
+-	if (!packet->giaddr.s_addr)
+-		packet->giaddr = ip->addresses[0];
++	if (!packet->giaddr.s_addr) {
++		if (enable_support_for_dual_tor) {
++			packet->giaddr = uplink->addresses[0];
++		} else {
++			packet->giaddr = ip->addresses[0];
++		}
++	}
++
+ 	if (packet->hops < max_hop_count)
+ 		packet->hops = packet->hops + 1;
+ 	else
+@@ -1231,7 +1294,7 @@ find_interface_by_agent_option(struct dhcp_packet *packet,
+ 	int i = 0;
+ 	u_int8_t *circuit_id = 0;
+ 	unsigned circuit_id_len = 0;
+-	struct interface_info *ip;
++	struct interface_info *ip = NULL;
+ 
+ 	while (i < len) {
+ 		/* If the next agent option overflows the end of the
+@@ -1262,14 +1325,20 @@ find_interface_by_agent_option(struct dhcp_packet *packet,
+ 		return (-1);
+ 	}
+ 
+-	/* Scan the interface list looking for an interface whose
+-	   name matches the one specified in circuit_id. */
+-
+-	for (ip = interfaces; ip; ip = ip->next) {
+-		if (ip->circuit_id &&
+-		    ip->circuit_id_len == circuit_id_len &&
+-		    !memcmp(ip->circuit_id, circuit_id, circuit_id_len))
+-			break;
++	if (enable_support_for_dual_tor) {
```

Comment:
is this piece of code an optimization? The interface would still be found, right?

Reply:
the code is not necessary there, updated